python/cudf/cudf/core/dtype_matching.py

```diff
@@ -0,0 +1,259 @@
+        itertools.chain.from_iterable(ancestors[node] for node in subnodes)
+    )
+
+
+def promotion_table(graph: Dict[T, Set[T]]) -> Dict[Tuple[T, T], T]:
+    """
+    Return a type promotion table for all pairs of types in the
+    provided partial semi-lattice.
+
+    Raises RuntimeError if the input is not a DAG
+
+    Raises ValueError if any pair does not have a unique least upper
+    bound.
+
+    The promotion for each pair in the input is the unique least upper
+    bound of the pair in the input. The absence of a pair in the
+    return value means that no unique upper bound exists, so there is
+    no safe promotion route for that pair.
+
+    Arrows in the graph should point from child to parent.
+    """
+    successors = ancestors(graph)
+    table = dict()
+    for pair in itertools.product(graph, graph):
+        lca = least_common_ancestors(*pair, ancestors=successors)
+        if len(lca) > 1:
+            # Malformed input
+            raise ValueError("Provided DAG is not a partial semi-lattice")
+        try:
+            # Is there a unique least upper bound?
+            (table[pair],) = lca
+        except ValueError:
+            # No, no safe promotion route
+            pass
+    return table
+
+
+@functools.cache
+def numeric_promotions() -> Dict[Tuple[np.dtype, np.dtype], np.dtype]:
+    """
+    Return a type promotion table for all pairs of non-decimal numeric types.
+
+    If a pair exists as a key in the return value, the value is the
+    safe type to promote to. If a pair does not exist, there is no
+    safe promotion route.
+    """
+    # Direct promotion routes for integer types
+    # +----+    +-----+    +-----+    +-----+
+    # | i8 |--->| i16 |--->| i32 |--->| i64 |
+    # +----+    +--^--+    +--^--+    +--^--+
+    #             /          /          /
+    #         /---'      /---'      /---'
+    #     /---'      /---'      /---'
+    # +--+-+    +---+-+    +---+-+    +-----+
+    # | u8 |--->| u16 |--->| u32 |--->| u64 |
+    # +----+    +-----+    +-----+    +-----+
+    int_lattice: Dict[np.dtype, Set[np.dtype]] = {
+        np.dtype("uint8"): {np.dtype("uint16"), np.dtype("int16")},
+        np.dtype("int8"): {np.dtype("int16")},
+        np.dtype("uint16"): {np.dtype("uint32"), np.dtype("int32")},
+        np.dtype("int16"): {np.dtype("int32")},
+        np.dtype("uint32"): {np.dtype("uint64"), np.dtype("int64")},
+        np.dtype("int32"): {np.dtype("int64")},
+        np.dtype("uint64"): set(),
+        np.dtype("int64"): set(),
+    }
+    # Direct promotion routes for float types
+    # +-----+    +-----+    +-----+
+    # | f16 |--->| f32 |--->| f64 |
+    # +-----+    +-----+    +-----+
+    float_lattice: Dict[np.dtype, Set[np.dtype]] = {
+        np.dtype("float16"): {np.dtype("float32")},
+        np.dtype("float32"): {np.dtype("float64")},
+        np.dtype("float64"): set(),
+    }
+    return {**promotion_table(int_lattice), **promotion_table(float_lattice)}
```

Comment:
The results here are small enough that we could just write out all the pairs, but this code is generic to handle any partial semi-lattice of promotion rules.